keps/sig-node/4010-head-request-probes/README.md

```diff
@@ -0,0 +1,265 @@
+
+## Motivation
+
+The motivation behind this proposal is to enhance Kubernetes' readiness and liveness probes by supporting HTTP HEAD requests. While GET requests are commonly used, they might not always be suitable for certain scenarios, such as checking the availability of resources without fetching their content. The ability to use HEAD requests will enable more efficient health checks for specific use cases, particularly those involving customer-facing services. By extending the probe functionality, Kubernetes aims to provide a more versatile and customizable platform for managing container health.
+
+### Goals
+
+- Allow Kubernetes users to utilize HTTP HEAD requests for defining readiness and liveness probes.
+- Enhance the probe functionality to support more efficient and lightweight health checks.
+- Provide additional flexibility for customer-facing applications that require specific health check behaviors.
+
+### Non-Goals
+
+- Modifying existing HTTP GET-based probes.
+- Introducing new HTTP request methods beyond HEAD.
+
+## Proposal
+
+The proposal involves extending Kubernetes' readiness and liveness probes to support HTTP HEAD requests. This will involve modifying the Kubernetes API to allow users to specify the request method as HEAD for these probes. When a HEAD request is used, only the response headers will be retrieved, without the response body. This lightweight operation is ideal for checking the availability of resources without incurring unnecessary network traffic or consuming additional system resources.
+
+For example, a pod's readiness probe configuration can be updated as follows:
+
+```yaml
+readinessProbe:
+  httpHead:
+    headers:
+      - name: X-Custom-Header
+        value: some-value
+    queryParams:
+      - name: token
+        value: abc123
+    path: /health
+    port: 8080
+    scheme: HTTPS
+    timeoutSeconds: 3
+    periodSeconds: 5
+```
+
+In this example, the readiness probe is configured to use an HTTP HEAD request with custom headers, query parameters, and a specific path, port, and scheme.
+
+There is an [untested solution how this feature might be implemented](https://github.com/kubernetes/kubernetes/commit/e3d556f0920ff736fec5df73c42f6db93f436a37).  
+Note: this solution is also missing a feature gate, which will be required for the final implementation.
+
+### Risks and Mitigations
+
+Introducing support for HTTP HEAD requests in probes may pose the following risks:
+
+- Increased complexity: Adding support for HEAD requests may introduce additional complexity to the Kubernetes codebase. This risk can be mitigated through careful design and thorough testing.
+- Backward compatibility: Existing readiness and liveness probes that use GET requests may need to be updated to utilize the new HEAD functionality. Documentation and upgrade guides can help users transition smoothly.
```

Comment:
This doesn't sound like backward-compatibitibility issue.
If you want to use new functionality, sure you need to change your configs.

The most important thing (and this is what is backward compatibility) is to ensure that users who don't want the HEAD functionality aren't forced to any changes - everything has to remain working for them as is.

Reply:
You are totally right. I think we can drop this paragraph completely.

Reply:
please do